src/uct/tcp/tcp_ep.c

```diff
@@ -2067,17 +2073,28 @@ ucs_status_t uct_tcp_ep_flush(uct_ep_h tl_ep, unsigned flags,
     }
 
     status = uct_tcp_ep_check_tx_res(ep);
-    if (status == UCS_ERR_NO_RESOURCE) {
-        UCT_TL_EP_STAT_FLUSH_WAIT(&ep->super);
-        return UCS_ERR_NO_RESOURCE;
+    if (ucs_unlikely(status != UCS_OK)) {
+        if (status == UCS_ERR_NO_RESOURCE) {
+            return UCS_ERR_NO_RESOURCE;
+        }
+        return UCS_OK;
     }
 
-    if (ep->flags & UCT_TCP_EP_FLAG_PUT_TX_WAITING_ACK) {
-        status = uct_tcp_ep_put_comp_add(ep, comp, ep->tx.put_sn);
+    if (ep->last_acked_sn != ep->tx.sn) {
+        /* Decrement the sequence number to not consider the flush operation
```

Comment:
what does it mean?

Reply:
it fixes hang of the following loop:
```
do {
    uct_worker_progress(worker);
    uct_ep_flush(ep);
} while (status = UCS_INPROGRESS);
```

since EP is always has outstanding PUT sent by flush. So, we don't want to count them

Reply:
maybe insted call some internal function of tcp which will reuse the current sequence number?
it's confusing that sn is decremented

Reply:
I see that it is confusing, but it should almost duplicate PUT code. Is it ok?